src/InternPool.zig

```diff
@@ -507,54 +467,140 @@ pub const Cau = struct {
+/// (Specifically, 104225 / 140923)
+///
 /// `Nav.Repr` is the in-memory representation.
 pub const Nav = struct {
     /// The unqualified name of this `Nav`. Namespace lookups use this name, and error messages may use it.
     /// Additionally, extern `Nav`s (i.e. those whose value is an `extern`) use this name.
     name: NullTerminatedString,
     /// The fully-qualified name of this `Nav`.
     fqn: NullTerminatedString,
-    /// If the value of this `Nav` is resolved by semantic analysis, it is within this `Cau`.
-    /// If this is `.none`, then `status == .resolved` always.
-    analysis_owner: Cau.Index.Optional,
+    /// This field is populated iff this `Nav` is resolved by semantic analysis.
+    /// If this is `null`, then `status == .resolved` always.
+    analysis: ?struct {
+        namespace: NamespaceIndex,
+        zir_index: TrackedInst.Index,
+    },
     /// TODO: this is a hack! If #20663 isn't accepted, let's figure out something a bit better.
     is_usingnamespace: bool,
     status: union(enum) {
-        /// This `Nav` is pending semantic analysis through `analysis_owner`.
+        /// This `Nav` is pending semantic analysis.
         unresolved,
+        /// The type of this `Nav` is resolved; the value is queued for resolution.
+        type_resolved: struct {
+            type: InternPool.Index,
+            alignment: Alignment,
+            @"linksection": OptionalNullTerminatedString,
+            @"addrspace": std.builtin.AddressSpace,
+            is_const: bool,
+            is_threadlocal: bool,
+            /// This field is whether this `Nav` is a literal `extern` definition.
+            /// It does *not* tell you whether this might alias an extern fn (see #21027).
+            is_extern_decl: bool,
+        },
         /// The value of this `Nav` is resolved.
-        resolved: struct {
+        fully_resolved: struct {
             val: InternPool.Index,
             alignment: Alignment,
             @"linksection": OptionalNullTerminatedString,
             @"addrspace": std.builtin.AddressSpace,
         },
     },
 
-    /// Asserts that `status == .resolved`.
+    /// Asserts that `status != .unresolved`.
     pub fn typeOf(nav: Nav, ip: *const InternPool) InternPool.Index {
-        return ip.typeOf(nav.status.resolved.val);
+        return switch (nav.status) {
+            .unresolved => unreachable,
+            .type_resolved => |r| r.type,
+            .fully_resolved => |r| ip.typeOf(r.val),
+        };
+    }
+
+    /// Always returns `null` for `status == .type_resolved`. This function is inteded
+    /// to be used by code generation, since semantic analysis will ensure that any `Nav`
+    /// which is potentially `extern` is fully resolved.
+    /// Asserts that `status != .unresolved`.
+    pub fn getExtern(nav: Nav, ip: *const InternPool) ?Key.Extern {
+        return switch (nav.status) {
+            .unresolved => unreachable,
+            .type_resolved => null,
```

Comment:
this should be 
```zig
.type_resolved => unreachable,
```
because otherwise a bug in the compiler will, instead of crashing here, treat an extern as if it were not an extern.

right?

Reply:
Nope -- this is intentional, but a little subtle. The background is that in general, when lowering a **pointer** to a `Nav`, backends don't care about the `Nav`'s value -- that's set up when you `updateNav` -- but they *might* care about the type, and whether it's an extern. So, Sema is set up in such a way that it will always immediately fully resolve any `Nav` which might be an extern (that is, whose value might have key `.@"extern"`). Here's an excerpt from a commit message which explains it (although the original message had an important typo at the start which is fixed here):

> In general, analyzing ZIR `decl_ref` will only analyze `nav_ty` of the
> corresponding `Nav`. There are two exceptions to this. If the
> declaration is an `extern` declaration, then we immediately ensure the
> `Nav` value is resolved (which doesn't actually require any more
> analysis, since such a declaration has no value body anyway).
> Additionally, if the resolved type has type tag `.@"fn"`, we again
> immediately resolve the `Nav` value. The latter restriction is in place
> for two reasons:
> 
> * Functions are special, in that their externs are allowed to trivially
>   alias; i.e. with a declaration `extern fn foo(...)`, you can write
>   `const bar = foo;`. This is not allowed for non-function externs, and
>   it means that function types are the only place where it is possible
>   for a declaration `Nav` to have a `.@"extern"` value without actually
>   being declared `extern`. We need to identify this situation
>   immediately so that the `decl_ref` can create a pointer to the *real*
>   extern `Nav`, not this alias.
> * In certain situations, such as taking a pointer to a `Nav`, Sema needs
>   to queue analysis of a runtime function if the value is a function. To
>   do this, the function value needs to be known, so we need to resolve
>   the value immediately upon `&foo` where `foo` is a function.
> 
> This restriction is simple to codify into the eventual language
> specification, and doesn't limit the utility of this feature in
> practice.

This `getExtern` function should generally only be called from codegen/link, since during semantic analysis, there are certain times where the situation you point out could indeed come up. The idea is that when lowering e.g. a pointer to `some_nav`, the linker code can do `if (some_nav.getExtern(ip)) |e| ...` in the event that it needs special handling to emit a pointer to an extern.

Reply:
> If you call it from the backend, it's supposed to be fully resolved.

No, this is not accurate. If you call this from the backend, we have the guarantee that **if it would have value `.@"extern"`, then it is resolved**. It may *not* be resolved if we know for certain that it *isn't* an extern -- which, in almost all cases, we do know.

> What's an example of when it would be correct code to branch on the optional, and the control flow takes the `.type_resolved => null` branch?

This happens all the time: there's an opportunity for it any time codegen lowers a function which includes a `decl_ref` of something that's not extern. For instance, this code probably does it:
```zig
test {
    // When this test is codegenned, `foo` has its type resolved, but not its
    // value resolved (this specific fact is an implementation detail).
    // Because `foo` is not declared `extern`, and `u32` is not a function type,
    // we know it can't be an extern, so Sema knows that it is safe to queue
    // resolution of the *value* for later -- because it knows that `getExtern`
    // *would* return `null` if the value were resolved.
    // The linker wants to call `nav_of_foo.getExtern(ip)` to construct the value `&foo`.
    f(&foo);
}
fn f(ptr: *const anyopaque) void { ... }
// Once we `updateNav` on `foo` itself, *then* the backend is allowed to assume `foo` is fully resolved.
const foo: u32 = 123;
```

Reply:
OK I think I get it. Then we need a different function for getting an extern, not from the nav ref path, but from the update nav path, and that one should assert fully resolved.

Reply:
I don't think that's necessary, since in that case you also want to catch things like `.variable`, so you probably ought to just be writing:
```zig
switch (ip.indexToKey(nav.status.fully_resolved.val)) {
    .@"extern" => |e| {
        // extern
    }.
    .variable => |v| {
        // global mutable variable
    },
    .func => {
        // comptime function alias
    },
    else => {
        // global constant
    },
}
```

Reply:
Let's see... I have 4 callsites and 100% of them do `.?` on the return value, because they are already known to be externs.

Reply:
Hm, okay. If you'd find it useful, feel free to make a separate function in your wasm linker PR or similar.